File: ios/Sources/FlowbizOnsite/CanonicalJSON.swift
import Foundation

/// Canonical JSON writer for the wire `data` payload strings (SPEC §4/§5).
///
/// `JSONSerialization` renders doubles with up-to-17-digit decimal expansions
/// (`19.99` → `19.989999999999998`, `0.1` → `0.10000000000000001`) and escapes
/// forward slashes (`\/`); Android's org.json and the web tracker's
/// `JSON.stringify` do neither, so the produced wire strings diverge
/// byte-for-byte. This writer emits the canonical cross-platform form —
/// pinned to web `JSON.stringify` (the reference implementation):
///
/// - **numbers**: shortest round-trip digits, formatted with the ECMAScript
///   `Number::toString` layout rules — `19.99`, `0.1`, whole doubles without
///   a fraction part (`19.0` → `19`), fixed notation up to 21 digits
///   (`10000000`, not `1.0E7`), exponent form beyond (`1e+21`), `-0.0` → `0`
/// - **strings**: minimal escaping — only `"` `\` and control characters;
///   raw slashes, raw unicode
/// - **objects**: keys sorted by UTF-16 code units (deterministic output;
///   matches Kotlin `sorted()` and JS `Array.prototype.sort`)
///
/// Mirrored by the Kotlin `CanonicalJson`; both are pinned byte-for-byte by
/// `expected.data_canonical` in `shared/fixtures/`.
///
/// **Throws** on non-finite numbers (NaN/±Infinity) — the same contract as
/// the Kotlin serializer (org.json rejects non-finite doubles at tree-build
/// time). The SPEC §3 never-throw guarantee is applied at the public API
/// boundary (Slice 4), not here.
enum CanonicalJSON {

    struct WriteError: Error, CustomStringConvertible {
        let description: String
        init(_ description: String) { self.description = description }
    }

    /// Renders a JSON tree (`String` / `NSNumber` / `Bool` / `Int` / `Double`
    /// / `NSNull` / `[Any]` / `[String: Any]`) as a compact canonical string.
    static func render(_ value: Any) throws -> String {
        var out = ""
        try write(value, into: &out)
        return out
    }

    private static func write(_ value: Any, into out: inout String) throws {
        switch value {
        case let string as String:
            writeString(string, into: &out)
        case let number as NSNumber:
            if CFGetTypeID(number) == CFBooleanGetTypeID() {
                out += number.boolValue ? "true" : "false"
            } else {
                out += try numberToken(number)
            }
        case is NSNull:
            out += "null"
        case let object as [String: Any]:
            out += "{"
            let keys = object.keys.sorted { $0.utf16.lexicographicallyPrecedes($1.utf16) }
            for (index, key) in keys.enumerated() {
                if index > 0 { out += "," }
                writeString(key, into: &out)
                out += ":"
                try write(object[key]!, into: &out)
            }
            out += "}"
        case let array as [Any]:
            out += "["
            for (index, element) in array.enumerated() {
                if index > 0 { out += "," }
                try write(element, into: &out)
            }
            out += "]"
        default:
            throw WriteError("unsupported JSON value of type \(type(of: value))")
        }
    }

    // MARK: - Strings

    /// Minimal escaping, matching `JSON.stringify`: `"` and `\` plus control
    /// characters; everything else (slashes, unicode) is emitted raw.
    private static func writeString(_ string: String, into out: inout String) {
        out += "\""
        for scalar in string.unicodeScalars {
            switch scalar {
            case "\"": out += "\\\""
            case "\\": out += "\\\\"
            case "\u{08}": out += "\\b"
            case "\u{09}": out += "\\t"
            case "\u{0A}": out += "\\n"
            case "\u{0C}": out += "\\f"
            case "\u{0D}": out += "\\r"
            default:
                if scalar.value < 0x20 {
                    out += String(format: "\\u%04x", scalar.value)
                } else {
                    out.unicodeScalars.append(scalar)
                }
            }
        }
        out += "\""
    }

    // MARK: - Numbers

    private static func numberToken(_ number: NSNumber) throws -> String {
        switch String(cString: number.objCType) {
        case "f", "d":
            return try doubleToken(number.doubleValue)
        case "Q":
            return "\(number.uint64Value)"
        default:
            return "\(number.int64Value)"
        }
    }

    /// ECMAScript `Number::toString(10)` rendering of a finite double, built
    /// from Swift's shortest-round-trip `"\(Double)"` digits.
    static func doubleToken(_ value: Double) throws -> String {
        guard value.isFinite else {
            throw WriteError("JSON does not allow non-finite numbers (\(value))")
        }
        if value == 0 { return "0" } // covers -0.0 → "0" (JSON.stringify(-0))

        // Parse the shortest representation, e.g. "19.99", "10000000.0",
        // "1e+21", "1e-07", into sign + digit string + decimal exponent.
        var repr = Substring("\(value)")
        var sign = ""
        if repr.first == "-" {
            sign = "-"
            repr = repr.dropFirst()
        }
        var mantissa = repr
        var exp10 = 0
        if let eIndex = repr.firstIndex(where: { $0 == "e" || $0 == "E" }) {
            mantissa = repr[..<eIndex]
            exp10 = Int(repr[repr.index(after: eIndex)...]) ?? 0
        }
        var digits: [Character]
        var pointPosition: Int
        if let dotIndex = mantissa.firstIndex(of: ".") {
            digits = Array(mantissa[..<dotIndex]) + Array(mantissa[mantissa.index(after: dotIndex)...])
            pointPosition = mantissa.distance(from: mantissa.startIndex, to: dotIndex)
        } else {
            digits = Array(mantissa)
            pointPosition = digits.count
        }
        // `n` per ECMA-262 Number::toString: value == 0.digits × 10^n.
        var n = pointPosition + exp10
        var start = 0
        while start < digits.count - 1 && digits[start] == "0" {
            start += 1
            n -= 1
        }
        digits.removeFirst(start)
        while digits.count > 1 && digits.last == "0" {
            digits.removeLast()
        }
        let k = digits.count
        let digitString = String(digits)

        if k <= n && n <= 21 {
            return sign + digitString + String(repeating: "0", count: n - k)
        }
        if 0 < n && n <= 21 {
            let split = digitString.index(digitString.startIndex, offsetBy: n)
            return sign + digitString[..<split] + "." + digitString[split...]
        }
        if -6 < n && n <= 0 {
            return sign + "0." + String(repeating: "0", count: -n) + digitString
        }
        let exponent = n - 1
        let head = k == 1 ? digitString : "\(digitString.first!)." + digitString.dropFirst()
        return sign + head + "e" + (exponent >= 0 ? "+" : "-") + String(abs(exponent))
    }
}

File: android/sdk/src/main/kotlin/br/com/flowbiz/onsite/CanonicalJson.kt
package br.com.flowbiz.onsite

import org.json.JSONArray
import org.json.JSONObject
import java.util.Locale
import kotlin.math.abs

/**
 * Canonical JSON writer for the wire `data` payload strings (SPEC §4/§5).
 *
 * org.json's own `toString()` diverges from the web tracker's
 * `JSON.stringify` (the reference implementation): `1.0E7`-style exponents
 * for large doubles (JVM), `-0` for negative zero, `\u2014`-style escaping
 * of U+2000–U+20FF characters, and `<\/` slash escaping. This writer emits
 * the canonical cross-platform form instead:
 *
 * - **numbers**: `Double.toString` digits formatted with the ECMAScript
 *   `Number::toString` layout rules — `19.99`, `0.1`, whole doubles without
 *   a fraction part (`19.0` → `19`), fixed notation up to 21 digits
 *   (`10000000`, not `1.0E7`), exponent form beyond (`1e+21`), `-0.0` → `0`.
 *   Caveat: on JDK ≤ 18 (Android included) `Double.toString` is *not*
 *   guaranteed shortest-round-trip (JDK-4511638, fixed in JDK 19), so a few
 *   extreme magnitudes carry extra digits vs JS/Swift — e.g. `1e23` renders
 *   `9.999999999999999e+22` (JS: `1e+23`) and `5e-324` renders `4.9e-324`
 *   (JS: `5e-324`). Both parse back to the identical double; realistic
 *   payload values (prices, quantities) are unaffected. Pinned by
 *   `CanonicalJsonNumberTest`.
 * - **strings**: minimal escaping — only `"` `\` and control characters;
 *   raw slashes, raw unicode
 * - **objects**: keys sorted by UTF-16 code units (deterministic output;
 *   matches Swift's UTF-16 sort and JS `Array.prototype.sort`)
 *
 * Mirrored by the Swift `CanonicalJSON`; both are pinned byte-for-byte by
 * `expected.data_canonical` in `shared/fixtures/`.
 *
 * **Throws** on non-finite numbers (NaN/±Infinity) — the same contract as
 * the Swift serializer (org.json already rejects non-finite doubles with a
 * `JSONException` at tree-build time; [numberToJson] re-checks). The SPEC §3
 * never-throw guarantee is applied at the public API boundary (Slice 4),
 * not here.
 */
internal object CanonicalJson {

    /** Renders an org.json tree as a compact canonical string. */
    fun render(value: Any?): String = buildString { appendValue(this, value) }

    private fun appendValue(out: StringBuilder, value: Any?) {
        when {
            value == null || value === JSONObject.NULL -> out.append("null")
            value is String -> appendString(out, value)
            value is Boolean -> out.append(if (value) "true" else "false")
            value is Int || value is Long -> out.append(value.toString())
            // Double/Float and org.json's parsed BigDecimals: all treated as
            // doubles, matching iOS (`JSONValue.number(Double)`) and JS.
            value is Number -> out.append(numberToJson(value.toDouble()))
            value is JSONObject -> {
                out.append('{')
                val keys = buildList {
                    val iterator = value.keys()
                    while (iterator.hasNext()) add(iterator.next() as String)
                }.sorted()
                keys.forEachIndexed { index, key ->
                    if (index > 0) out.append(',')
                    appendString(out, key)
                    out.append(':')
                    appendValue(out, value.get(key))
                }
                out.append('}')
            }
            value is JSONArray -> {
                out.append('[')
                for (index in 0 until value.length()) {
                    if (index > 0) out.append(',')
                    appendValue(out, value.get(index))
                }
                out.append(']')
            }
            else -> throw IllegalArgumentException("unsupported JSON value of type ${value::class.java.name}")
        }
    }

    // --- Strings ---

    /**
     * Minimal escaping, matching `JSON.stringify`: `"` and `\` plus control
     * characters (and lone surrogates, per well-formed `JSON.stringify`);
     * everything else — slashes, unicode — is emitted raw.
     */
    private fun appendString(out: StringBuilder, value: String) {
        out.append('"')
        var i = 0
        while (i < value.length) {
            val c = value[i]
            when {
                c == '"' -> out.append("\\\"")
                c == '\\' -> out.append("\\\\")
                c == '\b' -> out.append("\\b")
                c == '\t' -> out.append("\\t")
                c == '\n' -> out.append("\\n")
                c == '\u000C' -> out.append("\\f")
                c == '\r' -> out.append("\\r")
                c < ' ' -> out.append(String.format(Locale.ROOT, "\\u%04x", c.code))
                c.isHighSurrogate() && i + 1 < value.length && value[i + 1].isLowSurrogate() -> {
                    out.append(c).append(value[i + 1])
                    i++
                }
                c.isSurrogate() -> out.append(String.format(Locale.ROOT, "\\u%04x", c.code))
                else -> out.append(c)
            }
            i++
        }
        out.append('"')
    }

    // --- Numbers ---

    /**
     * ECMAScript `Number::toString(10)` rendering of a finite double, built
     * from Java's `Double.toString` digits (shortest round-trip only from
     * JDK 19; see the class doc for the JDK-4511638 caveat).
     */
    fun numberToJson(value: Double): String {
        require(!value.isNaN() && !value.isInfinite()) {
            "JSON does not allow non-finite numbers ($value)"
        }
        if (value == 0.0) return "0" // covers -0.0 → "0" (JSON.stringify(-0))

        // Parse the shortest representation, e.g. "19.99", "19.0", "1.0E21",
        // "1.0E-7", into sign + digit string + decimal exponent.
        var repr = value.toString()
        var sign = ""
        if (repr.startsWith("-")) {
            sign = "-"
            repr = repr.substring(1)
        }
        val eIndex = repr.indexOfFirst { it == 'e' || it == 'E' }
        var mantissa = repr
        var exp10 = 0
        if (eIndex >= 0) {
            mantissa = repr.substring(0, eIndex)
            exp10 = repr.substring(eIndex + 1).toInt()
        }
        val dotIndex = mantissa.indexOf('.')
        val rawDigits = if (dotIndex >= 0) mantissa.removeRange(dotIndex, dotIndex + 1) else mantissa
        // `n` per ECMA-262 Number::toString: value == 0.digits × 10^n.
        var n = (if (dotIndex >= 0) dotIndex else mantissa.length) + exp10
        var start = 0
        while (start < rawDigits.length - 1 && rawDigits[start] == '0') {
            start++
            n--
        }
        var end = rawDigits.length
        while (end - start > 1 && rawDigits[end - 1] == '0') end--
        val digits = rawDigits.substring(start, end)
        val k = digits.length

        return when {
            n in k..21 -> sign + digits + "0".repeat(n - k)
            n in 1..21 -> sign + digits.substring(0, n) + "." + digits.substring(n)
            n in -5..0 -> sign + "0." + "0".repeat(-n) + digits
            else -> {
                val exponent = n - 1
                val head = if (k == 1) digits else digits[0] + "." + digits.substring(1)
                sign + head + "e" + (if (exponent >= 0) "+" else "-") + abs(exponent)
            }
        }
    }
}
